src/main/scala/is/hail/sparkextras/ContextRDD.scala

```diff
@@ -0,0 +1,159 @@
+import is.hail.expr.types._
+import is.hail.utils._
+
+import org.apache.spark._
+import org.apache.spark.rdd._
+import org.apache.spark.storage._
+import scala.reflect.ClassTag
+
+object ContextRDD {
+  def empty[C <: AutoCloseable, T: ClassTag](
+    sc: SparkContext,
+    mkc: () => C
+  ): ContextRDD[C, T] =
+    new ContextRDD(sc.emptyRDD[C => Iterator[T]], mkc)
+
+  // this one weird trick permits the caller to specify T without C
+  sealed trait Empty[T] {
+    def apply[C <: AutoCloseable](
+      sc: SparkContext,
+      mkc: () => C
+    )(implicit tct: ClassTag[T]
+    ): ContextRDD[C, T] = empty[C, T](sc, mkc)
+  }
+  private[this] object emptyInstance extends Empty[Nothing]
+  def empty[T] = emptyInstance.asInstanceOf[Empty[T]]
+
+  def union[C <: AutoCloseable, T: ClassTag](
+    sc: SparkContext,
+    xs: Seq[ContextRDD[C, T]]
+  ): ContextRDD[C, T] = {
+    assert(xs.length > 1)
+    val mkc = xs(0).mkc
+    assert(xs.tail.forall(_ == mkc))
+    union(sc, xs, mkc)
+  }
+
+  def union[C <: AutoCloseable, T: ClassTag](
+    sc: SparkContext,
+    xs: Seq[ContextRDD[C, T]],
+    mkc: () => C
+  ): ContextRDD[C, T] =
+    new ContextRDD(sc.union(xs.map(_.rdd)), mkc)
+
+  def weaken[C <: AutoCloseable, T: ClassTag](
+    rdd: RDD[T],
+    mkc: () => C
+  ): ContextRDD[C, T] =
+    new ContextRDD(rdd.mapPartitions(it => Iterator.single((ctx: C) => it)), mkc)
+
+  type ElementType[C, T] = C => Iterator[T]
+}
+
+class ContextRDD[C <: AutoCloseable, T: ClassTag](
+  val rdd: RDD[C => Iterator[T]],
+  val mkc: () => C
+) extends Serializable {
+  def run: RDD[T] =
+    rdd.mapPartitions { part => using(mkc()) { cc => part.flatMap(_(cc)) } }
+
+  private[this] def inCtx[U: ClassTag](
+    f: C => Iterator[U]
+  ): Iterator[C => Iterator[U]] = Iterator.single { (ctx: C) => f(ctx) }
+
+  private[this] def withoutContext[U: ClassTag](
+    f: Iterator[T] => Iterator[U]
+  ): ContextRDD[C, U] = new ContextRDD(rdd.map(_.andThen(f)), mkc)
+
+  def map[U: ClassTag](f: T => U): ContextRDD[C, U] =
+    withoutContext(_.map(f))
+
+  def filter(f: T => Boolean): ContextRDD[C, T] =
+    withoutContext(_.filter(f))
+
+  def flatMap[U: ClassTag](f: T => TraversableOnce[U]): ContextRDD[C, U] =
+    withoutContext(_.flatMap(f))
+
+  def mapPartitions[U: ClassTag](
+    f: Iterator[T] => Iterator[U],
+    preservesPartitioning: Boolean = false
+  ): ContextRDD[C, U] =
+    cmapPartitions((_, part) => f(part), preservesPartitioning)
+
+  def mapPartitionsWithIndex[U: ClassTag](
+    f: (Int, Iterator[T]) => Iterator[U]
+  ): ContextRDD[C, U] = cmapPartitionsWithIndex((i, _, part) => f(i, part))
+
+  def zipPartitions[U: ClassTag, V: ClassTag](
+    that: ContextRDD[C, U],
+    preservesPartitioning: Boolean = false
+  )(f: (Iterator[T], Iterator[U]) => Iterator[V]
+  ): ContextRDD[C, V] =
+    czipPartitions[U, V](that, preservesPartitioning)((_, l, r) => f(l, r))
+
+  private[this] def withContext[U: ClassTag](
+    f: (C, Iterator[T]) => Iterator[U]
```

Comment:
And this cmapPartitions?

Reply:
I think part of what is making this confusing is that the interface isn't quite sure if a one-context-per-partition invariant is being enforced. I'm not yet convinced that we always want that invariant. If we're leaving that possibility open, it might help to clarify the interface if we give a name to the per-partition iterators. Maybe "blocks" (like a block of code, which is contained in one context/scope)? Then there would be separate `mapPartitions` (operating at the Spark partition level) and `mapBlocks`.

Reply:
Yeah basically what @patrick-schultz said and my comment on `withoutContext`.